### PR TITLE
feat(api-idorslug): Rename Path paramaters to `project_id_or_slug`

### PR DESCRIPTION
**Pull request by iamrajjoshi**
I am working on making our APIs work with slug and id in our path parameters. Over the path month, we have updated the endpoints so they can support both parameters. As a final step, we are renaming the parameters from `{resource}.slug` to `{resource}.id_or_slug`. 

There will be a similar PR for `organization_slug`

### CODE REVIEW

**Comment by iamrajjoshi** [comment]
i think its complaining about frontend changes b/c i updated api-docs

File: src/sentry/api/bases/project.py [hunk 1]
```diff
@@ -123,42 +123,44 @@ def convert_args(
...
+                    )
                     .select_related("organization")
                     .prefetch_related("teams")
                     .get()
                 )
         except Project.DoesNotExist:
             try:
                 # Project may have been renamed
-                # This will only happen if the passed in project_slug is a slug and not an id
+                # This will only happen if the passed in project_id_or_slug is a slug and not an id
```
Comment (nhsiehgit) [inline src/sentry/api/bases/project.py:144]: :squint:

wait

Comment (iamrajjoshi) [reply src/sentry/api/bases/project.py:144]: 🥔

**Review by nhsiehgit — COMMENTED** [review]
hmmm....

Rather than a straight replace, can we roll this out in a chunks?
it's a little scary to go through _all_ the api's in one fell swoop like this...

File: api-docs/paths/releases/project-release-file.json [hunk 2]
```diff
@@ -198,9 +198,9 @@
         }
       },
       {
-        "name": "project_slug",
+        "name": "projproject_id_or_slugct_slug",
```
Comment (sentaur-athena) [inline api-docs/paths/releases/project-release-file.json:201]: typo

File: src/sentry/feedback/blueprints/api.md [hunk 3]
```diff
@@ -155,7 +155,7 @@ Retrieve a collection of feedback items.
     }
     ```
 
-## Feedback [/projects/<organization_slug>/<project_slug>/feedback/<feedback_id>/]
+## Feedback [/projects/<organization_slug>/<project_id_or_slug>/feedback/<feedback_id>/]
```
Comment (sentaur-athena) [inline src/sentry/feedback/blueprints/api.md:158]: oh wow! you did good.

File: src/sentry/api/bases/project.py (hunk 1 above)
Comment (sentaur-athena) [reply src/sentry/api/bases/project.py:144]: Let's face it. This comment will rename confusing till removal of feature flag 😅

File: src/sentry/api/bases/project.py [hunk 4]
```diff
@@ -123,42 +123,44 @@ def convert_args(
...
                 redirect = ProjectRedirect.objects.select_related("project")
                 if id_or_slug_path_params_enabled(
                     self.convert_args.__qualname__, str(organization_slug)
                 ):
                     redirect = redirect.get(
                         organization__slug__id_or_slug=organization_slug,
-                        redirect_slug=project_slug,
+                        redirect_slug=project_id_or_slug,
```
[the remark this answers is not in the document]
Comment (sentaur-athena) [reply src/sentry/api/bases/project.py:151]: No we're not confirming here. The only place we can be sure it's slug is in the `else`.

**Review by sentaur-athena — COMMENTED** [review]
Left a few comments. Looks good in general.

**Comment by sentaur-athena** [comment]
@nhsiehgit just to clarify Raj has rolled out the feature flag already little by little and they've been soaking in GA for a while. His current change is just renaming the variable the way it shows up in the docs. Do you think it makes sense to roll that out eventually? 🤔 I don't think it makes sense and it makes his work harder because he has to maintain two variable for docs purposes.

File: src/sentry/api/bases/project.py (hunk 1 above)
Comment (iamrajjoshi) [reply src/sentry/api/bases/project.py:144]: yeah, once i remove the function, i think i can clean this up

**Comment by nhsiehgit** [comment]
hmm
@sentaur-athena if you're comfortable with it i can leave it to you to approve 😅 
I don't think i have the context here to fully follow

**Comment by anonrig** [comment]
I might be missing some key information here. But for the sake of people who don't know the reasoning:

- why do we start to support both id and slug at the same time? 
- What is the performance implication of this change? 
- Does this mean we have to index both slug and id separately on the database with other fields' combinations? 
- What is the cost of operating these indexes?

**Comment by mitsuhiko** [comment]
@anonrig you will find most of those questions answered here: https://www.notion.so/sentry/Make-APIs-work-with-Slug-ID-cbaf68b883b843d499191e9cc4c116d3

The page is not fully capturing this and for the sake of public readers: slug renames need to be communicated through the Hybrid Cloud RPC system and are a pain.  We also persist slugs in the org level auth tokens. As a result an org slug rename currently breaks uploads and it requires synching through the hybrid cloud bridge.

We have decided a while back that it would be beneficial to support IDs in all cases so that we have a stable identifier in most communications that is unaffected by slug renames.

**Comment by sentaur-athena** [comment]
Our common objects are heavily queried by id already and indexed on id. The convert_arg changes that are happening now are basically getting the ID from API and turn it to slug so most backend except this one part would not change at all.

**Review by sentaur-athena — APPROVED** [review]

**Pull request merged**